### PR TITLE
Basic Caching Funtimes

### PR DESCRIPTION
**Pull request by designatednerd**
In This PR: 

- Adds Cache protocol and default implementation which is thread safe
- Adds `FileSystemDataCache` which allows you to store things in any folder on the filesystem
- Adds `InMemoryCache` which uses an underlying NSCache
- Adds a helper protocol to allow easy use of `Codable` when caching
- Adds a convenience `ImageCache` since that's like 98% of why I wanted to write this in the first place. 
- Tessssssssttttttssss

### CODE REVIEW

**Comment by designatednerd** [comment]
Note: Looks like circle wants to be dumb since the config's not on the main thing yet.

**Comment by csdodd** [comment]
Is there really no preexisting library that solves this problem?

**Comment by designatednerd** [comment]
@csdodd There are several caching libraries with different approaches like [SDWebImage](https://github.com/SDWebImage/SDWebImage) and [Hankeke](https://github.com/Haneke/Haneke), but 

- They're not generic - they're largely focused with caching images, and particularly for things where I'm downloading other kinds of files, I'd like to be able to do that more generically
- They're often [pretty complicated](https://github.com/SDWebImage/SDWebImage#architecture) for what they're trying to do. I wanted to strip this down to the studs and try to keep it relatively simple but flexible.

**Comment by designatednerd** [comment]
@csdodd better?

**Comment by csdodd** [comment]
Might make sense to delete `DemoApp` for now, since it is just empty.

File: CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift:26]: Can we license these images under the same license as this library?

Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift:26]: They're from wikipedia, I just grabbed them and put them in our repo so we'd have a consistent URL for them and make sure they don't change on us. I'll update the license to note these are covered by the same license as they are at Wikipedia.

**Comment by designatednerd** [comment]
@csdodd Planning to put some stuff in DemoApp in the next PR, along with a README.

File: CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift
Comment (csdodd) [reply CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift:26]: As well as referencing Fair Use from Wikipedia you'll need to add reference to this license `The Simpsons ™ & © Twentieth Century Fox Film Corp. All Rights Reserved.`

I'm sure it's not an issue, but we could go for public domain images instead.

Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCanTests/TestImageLoader.swift:26]: 🤔I could also just grab images from my Flickr page that I own the copyright to.

File: CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FilesystemPathHelper.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FilesystemPathHelper.swift]: For the sake of consistency this should really be `FileSystemPathHelper`

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: It looks like if you `init` this with a `folderURL` that is being used elsewhere then `clearAll` will delete not only the cached items, but everything else in the directory. This "gotcha" should either be documented, or ideally you could avoid this by actually caching files in a new directory inside `folderURL`

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift:86]: Just a note to make sure that this implementation detail is extremely important. This lib doesn't do any kind of HTTP Caching (which external users of this library may be expecting), so best to make this very clear in the README.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: Nit pick that applies to more location that just this.

You're using the nomenclature `folder` and `directory` - maybe not an issue with autocomplete, but there are benefits of using the same term throughout. Framework uses `directory` could be worth removing all references to  `folder`

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: Yeah good point.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift:86]: Yep, README is coming in the next PR, will make sure to note it

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: yeah good point.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (csdodd) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: Minor: Could `subdirectoryName` default to a value rather than having to be specified? The requirement of `subDirectoryName` is to avoid the issue of accidentally deleting files when clearing the cache, so it is in essence, an implementation detail of this library. By allowing `subDirectoryName` to be user-set you are leaking an abstraction.

**Comment by csdodd** [comment]
LGTM. Fixing the licensing of the included image assets and adding the README are important but can be done in a later PR.
I left a comment about a leaky abstraction but it's more an API design detail rather than a reason to block the merge.
CircleCI isn't building -- is it configured to run iOS builds and have we applied for free open source builds?

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift]: Ooh, interesting point 🤔might have it default to the class name - otherwise if you set up different caches it'd all use the same folder, and you'd run into the same issue of accidentally deleting shit you didn't mean to.

**Comment by designatednerd** [comment]
@csdodd Running (and passing) now - just improper configuration of the `xcodebuild` command.

I'll update images on master once this gets merged to stuff I own, then I'll update the tests in the next PR.

**Review by csdodd — APPROVED** [review]

**Comment by designatednerd** [comment]
@csdodd Addressed your most recent concerns, let me know if that works for you.

**Comment by csdodd** [comment]
Still approved

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift
Comment (yuliaveres) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift:64]: why completion is not fired on the queue  in this case ?

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift
Comment (yuliaveres) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift:97]: same about completion called not on the passed queue

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift
Comment (yuliaveres) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift:126]: and here

**Review by yuliaveres — COMMENTED** [review]
Sorry for late comments, just few questions.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/Cache.swift:64]: 🤦‍♀️good catch, both here and below.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (yuliaveres) [inline CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift:38]: this could be a `debugPrint` ?

File: CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift
Comment (yuliaveres) [inline CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift:17]: is this the best option for this sort of check ? As documentation says:

> The following methods are of limited utility. Attempting to predicate behavior based on the current state of the filesystem or a particular file on the filesystem is encouraging odd behavior in the face of filesystem race conditions. It's far better to attempt an operation (like loading a file or creating a directory) and handle the error gracefully than it is to try to figure out ahead of time whether the operation will succeed.

And I would say it is very true, we used to experience weird behaviour connected to this check in our last project.

File: CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/CacheTypes/FileSystemDataCache.swift:38]: this should use the logger

File: CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift
Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift:17]: Interesting - I don't think I've hit those kinds of issues even when I'm throwing things at it in testing at super-high speeds. I'll have to ping you later about what you're talking about with the prior project, I don't really remember what that was about.

Comment (designatednerd) [reply CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift:17]: Added notes in the methods around this. As I said, I haven't seen issues with it, but it's at least worth leaving a breadcrumb trail if people have issues with it.

Comment (yuliaveres) [reply CacheMeIfYouCan/CacheMeIfYouCan/Helpers/FileManagerHelper.swift:17]: 👍

**Pull request merged**
